nixos/modules/services/backup/tarsnap.nix

```diff
@@ -329,7 +345,50 @@ in
           CapabilityBoundingSet = [ "CAP_DAC_READ_SEARCH" ];
           PermissionsStartOnly = "true";
         };
-      }) gcfg.archives;
+      }) gcfg.archives) //
+
+      (mapAttrs' (name: cfg: nameValuePair "tarsnap-restore-${name}"{
+        description = "Tarsnap restore '${name}'";
+        requires    = [ "network-online.target" ];
+
+        path = [ pkgs.iputils pkgs.tarsnap pkgs.utillinux ];
+
+        ##
+        preStart = ''
+          while ! ping -q -c 1 v1-0-0-server.tarsnap.com &> /dev/null; do sleep 3; done
+        '';
+
+        script =
+        let
+          tarsnap = ''tarsnap --configfile "/etc/tarsnap/${name}.conf"'';
+          lastArchive = ''$(${tarsnap} --list-archives | sort | tail -1)'';
+          run = ''${tarsnap} -x -f "${lastArchive}" ${optionalString cfg.verbose "-v"}'';
+
+          in if (cfg.cachedir != null) then ''
+            mkdir -p ${cfg.cachedir}
+            chmod 0700 ${cfg.cachedir}
+
+            ( flock 9
```

Comment:
By the way, why one has to lock and write to files in cache dir? Seems like running `tarsnap --fsck` should be enough.

Reply:
As this module generates one service per archive, all with the same period, they all fire at the same time. The `flock` dance prevents locking issues while `--fsck` runs.

Reply:
I've traced back my email exchange with tarsnap developers and my possible thought process but I'm unsure that double lock is needed given that `tarsnap` locks the cache directory. Perhaps we could 

1. Lock `lockf` which is a lock `tarsnap` also uses;
2. Check if `firstrun` exists. This is our own lock that is used to force systemd services to order themselves;
3. If it doesn't exist create and lock it; we need to release our lock on `lockf` for `tarsnap` to do its job;
4. When it's done we relock `lockf` (I likely did this for consistency -- it shouldn't be needed);

Then we perform the operation itself, locking `firstrun` in process.

While `tarsnap --fsck` still starts `lockf` isn't locked; that's why we keep `firstrun` locked to avoid a race condition when other operation can start before it. `lockf` itself is locked so that we can check if `firstrun` exists to decide if `--fsck` is needed. Hence double locking -- this should have been commented in the code.

Perhaps this can be simplified somehow?

Reply:
To answer @mkaito -- this locking doesn't abort units, it properly orders them instead. I don't think it's a good idea to make a backup unit abort if another one is already running.